### PR TITLE
Add testing, and pass callback a true clone of properties which is more correct, less buggy\rTESTING=unit

### PR DESCRIPTION
**Pull request by scottmcmaster**
@danbosnichill , sorry, I missed a couple of late commits on this branch while task-switching yesterday. Please have a quick look.

### CODE REVIEW

File: lib/promoted/ruby/client/request_builder.rb [hunk 1]
```diff
@@ -125,7 +125,12 @@ def compact_one_insertion(insertion, compact_func)
 
           # Only need a copy if there are properties to compact.
           compact_insertion = insertion.dup
-          compact_insertion[:properties] = compact_func.call(insertion[:properties])
+
+          # Let the custom function work with a deep copy of the properties.
+          # There's really no way to work with a shallow copy and still be able
+          # to restore the correct insertion properties after a call to delivery.
+          new_props =  Marshal.load(Marshal.dump(insertion[:properties]))
+          compact_insertion[:properties] = compact_func.call(new_props)
```
Comment (danbosnichill) [inline lib/promoted/ruby/client/request_builder.rb:133]: We should chat in person about this.  One of our Typescript client callers passes a more complex object onto properties.  I don't know if simple deep copy will work for it.

File: spec/promoted/ruby/client_spec.rb [hunk 2]
```diff
@@ -500,16 +514,61 @@
...
       expect(deliver_resp).not_to be nil
       expect(deliver_resp.key?(:insertion)).to be true
 
       # No log request generated since there's no experiment and we delivered the request.
       expect(deliver_resp[:log_request]).to be nil
     end
+
+    it "can compact insertions with a subset of properties" do
+      to_compact_delivery_properties_func = Proc.new do |properties|
+          properties[:struct][:product].delete(:url)
```
Comment (danbosnichill) [inline spec/promoted/ruby/client_spec.rb:538]: I'll be curious to see what performance looks like eventually for the deep copy approach.  I know deep copy is safer.

**Review by danbosnichill — APPROVED** [review]

File: lib/promoted/ruby/client/request_builder.rb (hunk 1 above)
Comment (scottmcmaster) [reply lib/promoted/ruby/client/request_builder.rb:133]: I will enter an issue to revisit this for TS.

File: spec/promoted/ruby/client_spec.rb (hunk 2 above)
Comment (scottmcmaster) [reply spec/promoted/ruby/client_spec.rb:538]: Yes, per our discussion, I don't think the alternatives are great, unfortunately.

**Pull request merged**